### PR TITLE
Fix #4533: chained calls incorrectly wrapping enclosing implicit objects

### LINKED ISSUE #4533: Lexer: Chaining: Method chaining wraps enclosing object incorrectly
Input:
```coffee
x =
  a: f 3
  b: f 4
    .something()
```
Output:
```js
x = {
  a: f(3),
  b: f(4)
}.something();
```
But consider that
```coffee
x =
  a: 3 # <- no function call here
  b: f 4
    .something()
```
compiles to
```js
x = {
  a: 3,
  b: f(4).something()
};
```
I have a fix coming.

### PR DESCRIPTION
**Pull request by xixixao**
Fixes #4533.

I looked through the other open and closed method chaining issues and this doesn't seem to fix any of them. It's possible it fixes some other issues I'm not aware of (I kinda feel like it should change some behavior, but we don't have any existing tests covering it).

### CODE REVIEW

**Comment by GeoffreyBooth** [comment]
LGTM. Do you think it’s enough of a breaking-change risk that we should make this only on `2`? (I’m thinking probably not, but just wanted to make sure.)

**Comment by xixixao** [comment]
I think this should be fixed in `1` (alas it is a breaking change, so it should go into the "next" version in our brave no-semver world).

**Comment by GeoffreyBooth** [comment]
Can you please explain how this would be a breaking change?

File: test/formatting.coffee [hunk 1]
```diff
@@ -198,6 +201,14 @@ test "#1495, method call chaining", ->
   ).join ', '
   eq 'a, b, c', result
 
+test "method call chaining inside objects", ->
+  id = (x) -> x
+  result =
+    a: id 1
+    b: id {a: 42}
+      .a
+  eq 42, result.b
```
Comment (xixixao) [inline test/formatting.coffee]: Good catch, let me fix that.

**Comment by xixixao** [comment]
This is a breaking change in the sense that code which previously compiled now compiles to something else. Luckily the previous compilation was not really meaningful, so no one should have hopefully relied on it. Changes like these are rather common here.

**Comment by GeoffreyBooth** [comment]
FYI @xixixao I reviewed all the open issues. I put “Chaining:” in the titles for ones involving chaining or leading dots: https://github.com/jashkenas/coffeescript/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20Chaining

Do you mind reviewing those after you’re done with this? I bet most of them are related. I also put “Lexer:” in the title for anything that seemed like it had to do with the code in `lexer.coffee`, which you seem to be quite expert on.

**Comment by xixixao** [comment]
I've already done that, I'll put the summary here later. I'm definitely no expert in any part of the compiler.

**Comment by xixixao** [comment]
Rebased and fixed tests.

**Comment by GeoffreyBooth** [comment]
@lydell what do you think?

**Pull request merged**